File: AppSets-iOS/AppSets/com/app/appsets/server/model/Application.swift
//
//  Application.swift
//  AppSets
//
//  Created by caiju Xu on 2024/7/9.
//

import Foundation

class Application : ImSessionHolder, Bio, Codable {
    
    public static let BIO_ID_PREFIX = "APP-"
    var appId:String? = nil
    var iconUrl:String? = nil
    var website:String? = nil
    var updateTime:String? = nil
    var createTime:String? = nil
    var developerInfo:String? = nil
    var price:String? = nil
    var priceUnit:String? = nil
    var bannerUrl:String? = nil
    var createUid:String? = nil
    var updateUid:String? = nil
    var name:String? = nil
    var category:String? = nil
    var platforms:[Platform]? = nil
    
    var id: String {
        get{
            return "\(Application.BIO_ID_PREFIX)\(appId ?? UUID().uuidString)"
        }
    }
    
    var bioUrl: String? = nil
    
    var session: Session? = nil
    
    enum CodingKeys: CodingKey {
        case appId
        case iconUrl
        case website
        case updateTime
        case createTime
        case developerInfo
        case price
        case priceUnit
        case bannerUrl
        case createUid
        case updateUid
        case name
        case category
        case platforms
    }
    
}

struct Platform: Codable {
    var id:String? = nil
    var name:String? = nil
    var packageName:String? = nil
    var introduction:String? = nil
    var versionInfos:[VersionInfo]? = nil
}

struct VersionInfo: Codable {
    var id:String? = nil
    var versionIconUrl:String? = nil
    var versionBannerUrl:String? = nil
    var version:String? = nil
    var versionCode:String? = nil
    var changes:String? = nil
    var packageSize:String? = nil
    var privacyUrl:String? = nil
    var screenshotInfos:[ScreenshotInfo]? = nil
    var downloadInfo:[DownloadInfo]? = nil
}

struct ScreenshotInfo: Codable {
    var id:String? = nil
    var createUid:String? = nil
    var updateTime:String? = nil
    var createTime:String? = nil
    var type:String? = nil
    var contentType:String? = nil
    var url:String? = nil
}

struct DownloadInfo: Codable {
    var id:String? = nil
    var createUid:String? = nil
    var updateUid:String? = nil
    var createTime:String? = nil
    var updateTime:String? = nil
    var downloadTimes:String? = nil
    var url:String? = nil
}

File: AppSets-Android/appsets/src/main/kotlin/xcj/app/appsets/server/model/Application.kt
package xcj.app.appsets.server.model

import android.os.Parcelable
import kotlinx.parcelize.Parcelize
import xcj.app.appsets.im.Bio

@Parcelize
data class Application(
    val appId: String,
    var iconUrl: String? = null,
    val website: String? = null,
    val updateTime: String? = null,
    val createTime: String? = null,
    val developerInfo: String? = null,
    val price: String? = null,
    val priceUnit: String? = null,
    var bannerUrl: String? = null,
    val createUid: String? = null,
    val updateUid: String? = null,
    val name: String? = null,
    val category: String? = null,
    val platforms: List<AppPlatform>? = null,
) : Bio {

    override val bioId: String
        get() = "$BIO_ID_PREFIX${appId}"

    override val bioName: String?
        get() = name

    override var bioUrl: Any? = null

    var currentVisiblePlatformPosition: Int = -1

    fun platformVersionDownloadInfos(platformName: String?): List<DownloadInfo> {
        if (platformName.isNullOrEmpty()) {
            return emptyList()
        }
        if (platforms.isNullOrEmpty()) {
            return emptyList()
        }
        platforms.forEach { platform ->
            if (platform.name == platformName) {
                return if (platform.versionInfos.isNullOrEmpty()) {
                    emptyList()
                } else {
                    platform.versionInfos.flatMap {
                        val downloadInfos = it.downloadInfos
                        if (downloadInfos.isNullOrEmpty()) {
                            emptyList()
                        } else {
                            downloadInfos
                        }
                    }
                }
            }
        }
        return emptyList()
    }

    fun hasPlatformDownloadInfo(platformName: String?): Boolean {
        if (platformName.isNullOrEmpty()) {
            return false
        }
        if (platforms.isNullOrEmpty()) {
            return false
        }
        platforms.forEach { platform ->
            if (platform.name == platformName) {
                if (platform.versionInfos.isNullOrEmpty()) {
                    return false
                } else {
                    platform.versionInfos.forEach { versionInfo ->
                        if (!versionInfo.downloadInfos.isNullOrEmpty()) {
                            return true
                        }
                    }
                }
            }
        }
        return false
    }


    companion object {

        const val BIO_ID_PREFIX = "APP-"

        fun basic(id: String, name: String?, iconUrl: String?): Application {
            return Application(appId = id, name = name, iconUrl = iconUrl).apply {
                bioUrl = iconUrl
            }
        }
    }
}

@Parcelize
data class AppPlatform(
    val id: String? = null,
    val name: String? = null,
    val packageName: String? = null,
    val introduction: String? = null,
    val versionInfos: List<VersionInfo>? = null,
) : Parcelable


@Parcelize
data class VersionInfo(
    val id: String? = null,
    var versionIconUrl: String? = null,
    var versionBannerUrl: String? = null,
    val version: String? = null,
    val versionCode: String? = null,
    val changes: String? = null,
    val packageSize: String? = null,
    val privacyUrl: String? = null,
    val screenshotInfos: List<ScreenshotInfo>? = null,
    val downloadInfos: List<DownloadInfo>? = null,
) : Parcelable

@Parcelize
data class ScreenshotInfo(
    val id: String? = null,
    val createUid: String? = null,
    val updateUid: String? = null,
    val createTime: String? = null,
    val updateTime: String? = null,
    val type: String? = null,
    val contentType: String? = null,
    var url: String? = null,
) : Parcelable

@Parcelize
data class DownloadInfo(
    val id: String? = null,
    val createUid: String? = null,
    val updateUid: String? = null,
    val createTime: String? = null,
    val updateTime: String? = null,
    val downloadTimes: Int?,
    val url: String? = null,
    val description: String? = null,
    val architectures: List<String>? = null,
    val size: String? = null,
) : Parcelable
